Gems/OpenParticleSystem/Tools/OpenParticleSystemEditor/Window/Controls/GradientColorPickerWidget.cpp

```diff
@@ -0,0 +1,549 @@
+                    }
+                    OnGradientChanged();
+                });
+            menu.exec(QCursor::pos());
+        }
+        else
+        {
+            if (!RestrictedKey())
+            {
+                QMenu menu;
+                menu.addAction("Delete Key", this, [key, this]()
+                    {
+                        RemoveKey(key->GetStop());
+                    });
+                menu.exec(QCursor::pos());
+            }
+        }
+    }
+
+    void GradientColorPickerWidget::OnPickColor(GradientKey* key)
+    {
+        const QColor existingColor = key->GetStop().second;
+
+        const AZ::Color color = AzQtComponents::ColorPicker::getColor(
+            AzQtComponents::ColorPicker::Configuration::RGBA, AzQtComponents::fromQColor(existingColor), tr("Select Color"));
+        QColor selectedColor = AzQtComponents::ToQColor(color);
+        if (selectedColor != existingColor)
+        {
+            QGradientStop stop = key->GetStop();
+            stop.second = selectedColor;
+            key->SetStop(stop);
+            OnGradientChanged();
+        }
+    }
+
+    void GradientColorPickerWidget::SetSelectedKeyPosition(int location)
+    {
+        float loc = location / m_percentage;
+        GradientKey* key = GetSelectedKey();
+        if (key != nullptr)
+        {
+            key->SetTime(loc);
+            OnGradientChanged();
+        }
+    }
+    void GradientColorPickerWidget::SetDefaultLocationChangedCB(AZStd::function<void()> callback)
+    {
+        m_defaultLocationChangedCB = callback;
+    }
+    void GradientColorPickerWidget::SetLocationChangedCB(AZStd::function<void(short, QColor)> callback)
+    {
+        m_locationChangedCB = callback;
+    }
+
+    void GradientColorPickerWidget::SetGradientChangedCB(AZStd::function<void(QGradientStops)> callback)
+    {
+        m_gradientChangedCB = callback;
+    }
+
+    QRect GradientColorPickerWidget::GetRegion()
+    {
+        QPoint topLeft(0, height() - m_gradientSize.height());
+        QPoint bottomRight(width(), height());
+        return QRect(topLeft, bottomRight);
+    }
+
+    GradientColorPickerWidget::GradientKey* GradientColorPickerWidget::GetSelectedKey()
+    {
+        for (int i = 0; i < m_keys.count(); i++)
+        {
+            if (m_keys[i].IsSelected())
+            {
+                return &m_keys[i];
+            }
+        }
+        return nullptr;
+    }
+
+    void GradientColorPickerWidget::OnGradientChanged()
+    {
+        QGradientStops keys = GetKeys();
+        if (!signalsBlocked())
+        {
+            if (m_gradientChangedCB)
+            {
+                blockSignals(true);
+                m_gradientChangedCB(keys);
+                blockSignals(false);
+            }
+            m_gradientWidget->SetKeys((unsigned int)Gradients::HUE_RANGE, keys);
+            m_gradientWidget->update();
+            update();
+        }
+    }
+
+#pragma region GradientKeyFunctions
```

Comment:
#pragma region is a hint to certain IDEs that can collapse code, but they need to have a #pragma endregion to make it work. purhaps some editors will intuit this to mean to the end of the file, but I say remove it.